### PR TITLE
Deprecate internal metrics in favor of TorchMetrics

### PR DESCRIPTION
**Pull request by Borda**
This PR is resolving #4283

Still some open questions/tasks:
- [ ] shall it be hard removal or with deprecation warning to not breaking user experience (as it is now in this draft)
- [ ] shall be TorchMetrics `install` or `full_install` dependency

### CODE REVIEW

**Comment by Borda** [comment]
@rusty1s seems that this PR a slightly lowering the code coverage but it is quite understandable...

**Comment by Borda** [comment]
@rusty1s may I get your opinion on this suggestion/change as the PR description: :rabbit: 
- hard removal or with deprecation warning to not breaking user experience (as it is now in this draft)
- TorchMetrics install or full_install dependency

**Comment by rusty1s** [comment]
@Borda Yes, sorry for the delay :)

* I think `torchmetrics` should be a dependency of the `full_install`.
* I'm fine with a hard removal TBH since I doubt that these are functions that most PyG users are using. If we go for a deprecation warning, is there any chance we can make use of the in-house [`deprecated` warning](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/deprecation.py) rather than relying on the external dependency?

File: torch_geometric/utils/metric.py [hunk 1]
```diff
@@ -138,41 +58,14 @@ def f1_score(pred: Tensor, target: Tensor, num_classes: int) -> Tensor:
...
 
 
+@deprecated(tm_func.jaccard_index, deprecated_in="v2.0", remove_in="v3.0",
+            args_mapping={
+                "pred": "preds",
+                "batch": None,
+                "omitnans": None
+            })
 def mean_iou(pred: Tensor, target: Tensor, num_classes: int,
              batch: Optional[Tensor] = None, omitnans: bool = False) -> Tensor:
```
Comment (rusty1s) [inline torch_geometric/utils/metric.py:70]: It looks like the `batch` argument support will be dropped here (mean `iou` across all examples). This might be a problem.

Comment (Borda) [reply torch_geometric/utils/metric.py:70]: ok, then we can do the batch over the IoU... but shall we do it while you suggest dropping metrics completely?

Comment (rusty1s) [reply torch_geometric/utils/metric.py:70]: What do you mean with doing the batch over the IoU?

**Comment by Borda** [comment]
> I'm fine with a hard removal TBH since I doubt that these are functions that most PyG users are actively using.

this is something I also do not know that is why I started wth soft removal and can be dropped completely in net release... but happy to follow any your suggestion :)

> * If we go for a deprecation warning, is there any chance we can make use of the in-house [`deprecated` warning](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/deprecation.py) rather than relying on the external dependency?

sure I was using a deprecation package I know, but happy to use any according to your suggestion :)

**Comment by rusty1s** [comment]
Let's go for hard removal :)

**Comment by Borda** [comment]
> Let's go for hard removal :)

Done :+1:

File: examples/dgcnn_segmentation.py [hunk 2]
```diff
@@ -86,11 +86,8 @@ def test(loader):
     for data in loader:
         data = data.to(device)
         pred = model(data).argmax(dim=1)
-
-        i, u = i_and_u(pred, data.y, loader.dataset.num_classes, data.batch)
-        iou = i.cpu().to(torch.float) / u.cpu().to(torch.float)
-        iou[torch.isnan(iou)] = 1
-
+        iou = jaccard_index(pred, data.y,
```
Comment (Borda) [inline examples/dgcnn_segmentation.py:89]: @rusty1s so you say We shall implement the batch strategy here, in the example, right?

Comment (rusty1s) [reply examples/dgcnn_segmentation.py:89]: Let me verify the output in a mini-batch scenario (with a dynamic number of points per example) and come back to you (probably on Monday). This is my only concern. Otherwise the PR looks good to me (thank you very much).

Comment (rusty1s) [reply examples/dgcnn_segmentation.py:89]: What I mean by this is that I think we need to compute `jacard_index` with `reduction = None`, and then apply the reduction by ourselves, e.g., via `scatter_mean(iou, batch, dim=0).mean()`.

Comment (Borda) [reply examples/dgcnn_segmentation.py:89]: sure, because otherwise we can use also the `nn.Module` metrics version with `update` and `compute` :rabbit:

Comment (Borda) [reply examples/dgcnn_segmentation.py:89]: cc: @SkafteNicki

Comment (rusty1s) [reply examples/dgcnn_segmentation.py]: Updated the metric computation in the examples, in which IoU is now computed for each example individually and averaged afterwards. I think this leads to the correct computation now (please confirm).

I think there exists no alternative way to avoid the for-loop at the moment.

**Review by rusty1s — APPROVED** [review]

File: examples/dgcnn_segmentation.py [hunk 3]
```diff
@@ -80,24 +80,16 @@ def train():
...
-        iou = i.cpu().to(torch.float) / u.cpu().to(torch.float)
-        iou[torch.isnan(iou)] = 1
+        out = model(data).argmax(dim=1)
 
-        # Find and filter the relevant classes for each category.
-        for iou, category in zip(iou.unbind(), data.category.unbind()):
-            ious[category.item()].append(iou[y_mask[category]])
+        sizes = (data.ptr[1:] - data.ptr[:-1]).tolist()
+        for pred, y in zip(out.split(sizes), data.y.split(sizes)):
+            iou += float(jaccard_index(pred, y))
```
Comment (SkafteNicki) [inline examples/dgcnn_segmentation.py:90]: @Borda in v0.8 the num_classes will be required in `jaccard_index`
```suggestion
            iou += float(jaccard_index(pred, y, num_classes=loader.dataset.num_classes))
```

File: examples/point_transformer_segmentation.py [hunk 4]
```diff
@@ -197,24 +197,16 @@ def train():
...
-        iou = i.cpu().to(torch.float) / u.cpu().to(torch.float)
-        iou[torch.isnan(iou)] = 1
+        out = model(data.x, data.pos, data.batch).argmax(dim=1)
 
-        # Find and filter the relevant classes for each category.
-        for iou, category in zip(iou.unbind(), data.category.unbind()):
-            ious[category.item()].append(iou[y_mask[category]])
+        sizes = (data.ptr[1:] - data.ptr[:-1]).tolist()
+        for pred, y in zip(out.split(sizes), data.y.split(sizes)):
+            iou += float(jaccard_index(pred, y))
```
Comment (SkafteNicki) [inline examples/point_transformer_segmentation.py:207]:
```suggestion
            iou += float(jaccard_index(pred, y, num_classes=loader.dataset.num_classes))
```

File: examples/pointnet2_segmentation.py [hunk 5]
```diff
@@ -106,24 +106,16 @@ def train():
...
-        iou = i.cpu().to(torch.float) / u.cpu().to(torch.float)
-        iou[torch.isnan(iou)] = 1
+        out = model(data).argmax(dim=1)
 
-        # Find and filter the relevant classes for each category.
-        for iou, category in zip(iou.unbind(), data.category.unbind()):
-            ious[category.item()].append(iou[y_mask[category]])
+        sizes = (data.ptr[1:] - data.ptr[:-1]).tolist()
+        for pred, y in zip(out.split(sizes), data.y.split(sizes)):
+            iou += float(jaccard_index(pred, y))
```
Comment (SkafteNicki) [inline examples/pointnet2_segmentation.py:116]:
```suggestion
            iou += float(jaccard_index(pred, y, num_classes=loader.dataset.num_classes))
```

**Review by SkafteNicki — APPROVED** [review]

File: examples/dgcnn_segmentation.py (hunk 2 above)
Comment (rusty1s) [reply examples/dgcnn_segmentation.py]: This leads to:
`You have set 50 number of classes which is different from predicted (32) and target (3) number of classes`
